Plugin integration tests

AdamBJ opened the pull request:
This PR adds integration tests to the plugin

- Adds several integration tests (see `UmplePluginIntegrationTests.groovy`), the most notable of which is `generateAndCompileJava`. The test generates Java files from the `master.ump` file of two subprojects, and then compiles the generated Java. JUnit then checks that the files were generated and compiled correctly.
- In support of `generateAndCompileJava`, a test project is included in this PR (see `src/integrationTests/resources/TestProject`). As alluded to in the previous point, the project contains a root project and two subprojects. @vahdat-ab, it should be possible to re-use the structure of `TestProject` to build the umple compiler jar. Imagine each of the dependency projects (`UmpleParser` etc) as subprojects. 
- Unit tests and integration tests can be run independently. Run `gradle test` to run the unit tests and `gradle integrationTest` to run the integration tests
- The unit tests and integration test are now dependencies of the build task, so whenever the plugin is built the unit and integration tests are run
- Modified build.gradle to prevent pathing issues

closes #9. At some point we'll need to address #7 and #8, but it's not urgent.

AdamBJ committed:
Adding integration tests
- integrationTest task is now part of the build (run after unit tests (i.e. the test task))
- added some basic integration tests. More will be added in another commit

AdamBJ committed:
Added integration tests

- Unit tests and integration tests can be run independently. Run `gradle test` to run the unit tests and `gradle integrationTest` to run the integration test.ump
- The unit tests and integration test are now dependencies of the build task, so whenever the plugin is build the unit and integration tests are run
- Modified build.gradle to prevent pathing issues

AdamBJ commented:
@vahdat-ab @Nava2, I've updated the plugin on the plugin portal. You should be able to clone this repo, switch to the `Plugin_Integration_Tests` branch, and run `gradle build` in the `umple.gradle` directory to see everything in action.

Nava2 commented on build.gradle:
```diff
@@ -24,18 +24,52 @@ buildscript {
+  the integration test task runs fully every time we invoke it.
+*/
+project.integrationTest {
+    outputs.upToDateWhen { false }
```
This mechanism could be useful for the umple parts.

Nava2 commented on src/integrationTest/groovy/UmplePluginIntegrationTests.groovy:
```diff
@@ -0,0 +1,88 @@
+import static org.junit.Assert.*
+import org.gradle.testfixtures.ProjectBuilder
+import org.gradle.api.Project
+import org.gradle.tooling.* //GradleConnector, ProjectConnection, BuildLauncher
```
This should probably be fixed to remove the comment

Nava2 commented on src/integrationTest/groovy/UmplePluginIntegrationTests.groovy:
```diff
@@ -0,0 +1,88 @@
+    private static Project testProject
+    private static final DIRECTORIES_TO_CHECK = 
+    [
+        "src${File.separator}integrationTest${File.separator}resources${File.separator}TestProject${File.separator}subproj${File.separator}build${File.separator}classes${File.separator}generatedSource",
```
You should refactor these paths to have a `root`, i.e. `src/integrationTest/resources`. Then each array value is just the "subpath" for each. Additionally, I think using `/` should be okay here. Java converts the `/` to the separator anyway.

AdamBJ replied:
How about this?

```
private static final DIRECTORIES_ROOT = "src/integrationTest/resources/TestProject/"
    private static final DIRECTORIES_TO_CHECK = 
    [
        DIRECTORIES_ROOT + "subproj/build/classes/generatedSource",
        DIRECTORIES_ROOT + "subproj/generated/java",
        DIRECTORIES_ROOT + "sub2/build/classes/generatedSource",
        DIRECTORIES_ROOT + "sub2/generated/java"
    ] 
```

If the array just contains the subpaths for each directory we'd need to append the root before using any of the values in the array - that's easy to forget to do. We'd also be appending strings unnecessarily (rather than appending the root just once in the array we'd be appending the root and the subpath anywhere we wanted to use the array values)

Nava2 replied:
Use your judgement, definitely better this way than the original. 👍

Nava2 commented on src/integrationTest/groovy/UmplePluginIntegrationTests.groovy:
```diff
@@ -0,0 +1,88 @@
+
+    @Test
+    public void applyUmplePluginToProject() {      
+        assertTrue('Umple plugin was not applied', testProject.getPlugins().hasPlugin('umple.gradle.plugin'))
```
Neat, this mechanism work for Java too?

AdamBJ replied:
Yup! All you'd need to do to make the entire `UmplePluginIntegrationTests` file work in Java is add `;`s and change the way the `DIRECTORIES_TO_CHECK` list is initialized.

Nava2 replied:
No, I meant checking for the `java` plugin.

AdamBJ replied:
In that case I'm not sure. It'd depend on `org.gradle.api.Project`

Nava2 commented on src/integrationTest/groovy/UmplePluginIntegrationTests.groovy:
```diff
@@ -0,0 +1,88 @@
+            connection.close()
+        }
+        
+        for (String current : DIRECTORIES_TO_CHECK) {
```
This set of checks isn't good enough. We need to check that the files are properly generated. For example, check the main repository and rip out one of the simpler test cases. Then, check the tested code matches the expected. 

Unfortunately, this introduces some dependencies on Umple itself. I wonder if we should be extracting some of the test resources into a "fixture project." This side discussion is not relevant to this PR, though. Just copy a test case or two of content out and make sure the files are generated properly.

AdamBJ replied:
I'll keep the current checks for `.class` files, but I'll add content checks for generated Java. My thinking is that the `.class` files might change in the future if the way Java is compiled changes. Does that make sense?

Nava2 replied:
Sorry, but I don't see whe4re you check the class files. Where is this?

AdamBJ replied:
I don't explicitly check the contents of the class files. I check that the number that of class files is the number we expect (`if (count != 2)`), and that they aren't empty (`if (file.length() == 0) ...`).

AdamBJ replied:
You're talking about checking the contents of the files though, right? I think that's a good idea for the generated Java files, but I'm not so sure about the compile class files.

Nava2 replied:
You're not actually checking the class files for names or anything of that sort. It'd be good to check that, the content of a class file doesn't need to be checked.

Nava2 commented on src/integrationTest/resources/TestProject/build.gradle:
```diff
@@ -0,0 +1,34 @@
+  }
+  dependencies {
+    // Use the umple jar from the main umple.gradle project
+    classpath files('../../../../libs/umple-latest.jar')
```
Make the parent project copy this JAR somewhere local to this subproject. This path is too long and dangly.

Nava2 commented on src/integrationTest/resources/TestProject/build.gradle:
```diff
@@ -0,0 +1,34 @@
+ 
+ // Add required plugins and source sets to the sub projects
+subprojects{
+	apply plugin: "umple.gradle.plugin"
```
spacing

Nava2 commented on src/integrationTest/resources/TestProject/build.gradle:
```diff
@@ -0,0 +1,34 @@
+    generatedSource
+  }
+
+  compileGeneratedSourceJava.dependsOn('generateSource')
```
For now, I'm okay that this is manual. However, I would really rather this was done via our plugin.

Nava2 commented on src/main/groovy/cruise/umple/UmpleGradlePlugin.groovy:
```diff
@@ -12,10 +12,10 @@ class UmpleGradlePlugin implements Plugin<Project> {
-	private static final String DEFAULT_GENERATED_OUTPUT_PATH = "generated${File.separator}src${File.separator}java"
-	private static final String DEFAULT_UMPLE_FILE_PATH = "src${File.separator}master.ump"
+	private static final String DEFAULT_GENERATED_OUTPUT_PATH = "generated${File.separator}java"
+	private static final String DEFAULT_UMPLE_FILE_PATH = "src${File.separator}umple${File.separator}master.ump"
```
See my previous comments on the `File.separator`.

Nava2 commented on src/main/groovy/cruise/umple/UmpleGradlePlugin.groovy:
```diff
@@ -29,12 +29,14 @@ class UmpleGradlePlugin implements Plugin<Project> {
 			if(project.hasProperty(UMPLE_FILE_PATH))
 			{
-				m_umpleFilePath = project.getProperty(UMPLE_FILE_PATH)	
+				m_umpleFilePath += project.getProperty(UMPLE_FILE_PATH)	
```
This is direct changes of behaviour. Are the paths supposed to be relative to the project directory or can they be absolute? I would recommend using [`Path#relativize(Path)`](https://docs.oracle.com/javase/7/docs/api/java/nio/file/Path.html#relativize(java.nio.file.Path)) to make sure the paths behave properly and consistently.

AdamBJ replied:
I'm not sure what you're getting at. The only way I've been able to avoid problems is by making the path absolute. Ideally, I'd declare the default values on single lines before the plugin's `apply` method, but because I don't have access to `project.projectDir` until I'm inside the plugin's `apply` method I can't do that. So I have to make the default path values relative to the project directory.

Are you just suggesting I replace the string concatenation with calls to `relativize`? Or is this a bigger issue with how the paths stuff is implemented?

AdamBJ replied:
If you're talking about how the user specifies the paths, they should specify a path relative to the project directory. That's the way the plugin worked before and I wanted to keep it consistent.

AdamBJ replied:
@Nava2 I've got the PR fixed up except for this point. Can you be a bit more specific regarding what you want to see? I could use the `resolve` method to add two paths together instead of concatenating them as strings. Can't think of another way to include `Path`s, though.

Nava2 replied:
Are the paths relative to the project or absolute? Make sure it's relative to the project, and it works when the project structure changes.

AdamBJ replied:
The paths that the user specifies, as well as the default path constants, are relative to the project's main directory (i.e. the folder that contains `build.gradle`). If the project structure changes then the default values need to be changed to reflect the change.

Nava2 replied:
Are the paths relative to the project or absolute? Make sure it's relative to the project, and it works when the project structure changes.

AdamBJ replied:
I'm not sure what to make of that. Are you trying to tell me to ask fewer questions? I'm happy to take more ownership of the project, you seem to have a specific vision for how this project will turn out, though. I'm just trying to make sure we're both on the same page.

Nava2 replied:
> I'm not sure what to make of that. Are you trying to tell me to ask fewer questions? I'm happy to take more ownership of the project, you seem to have a specific vision for how this project will turn out, though. I'm just trying to make sure we're both on the same page.

Nooooo, oh God. I was.. using my phone during a break in a movie and I guess my connection crapped out. Oops. Didn't mean to send twice. 

Edit: Removed some information. 😊 

> The paths that the user specifies, as well as the default path constants, are relative to the project's main directory (i.e. the folder that contains build.gradle). If the project structure changes then the default values need to be changed to reflect the change.

As long as it works, that's what matters! If you can verify that, I'm happy.

Nava2 reviewed (changes requested):
* Definitely need to address the concern in #9 about figuring out how to test w/o the plugin hosted

While I made loads of comments, they are mostly smaller things -- except the path and UT stuff. Well done thus far, please make the requested changes, though.

AdamBJ committed:
Addressing Kevin's comments on PR #11. See https://github.com/umple/umple.gradle/pull/11#discussion_r103134167

Nava2 commented:
I think you've addressed all my comments, but I can't see any changes you said you were addressing? Can you add them to this PR? I think you just forgot to push.

AdamBJ committed:
Formatting, deleting accidentally committed files, minor bug fix

AdamBJ commented:
I haven't forgotten, just double-checking everything one last time

AdamBJ committed:
Added additional error handling to copy jar task in Umple Gradle plugin's build.gradle file

AdamBJ committed:
Formatting, adding more informative comment to UmpleGradlePlugin.groovy

AdamBJ commented:
@Nava2 Ok, the changes are pushed. Should address all your comments. The big change from when you last saw it is that I split the integration tests into two files, `Basic.groovy` and `CompilationAndGeneration.groovy`. `CompilationAndGeneration.groovy` now verifies file names for all files and file contents for `java` files.

Nava2 commented on build.gradle:
```diff
@@ -20,25 +20,83 @@ dependencies {
+// Run unit tests before integration tests
+integrationTest.mustRunAfter test
+
+task copyUmpleJar(type: Copy) {
```
Is this for testing? This handles copying the actual umple code?

AdamBJ replied:
Yes, it's for testing. It copies the Umple jar from it's location in the main Umple plugin project to the TestProject directory. That way we get rid of the "long and dangly path" in TestProject's `build.gradle` file.

Nava2 replied:
Okay, will this be necessary once we get a maven artifact published?

AdamBJ replied:
No. If we can just pull the umple jar from a maven repo the copy won't be necessary.

Nava2 replied:
Cool, works for me.

Nava2 commented on build.gradle:
```diff
@@ -59,27 +117,27 @@ uploadArchives {
+    compile files('./umple-latest.jar')
 }
 
 // The configuration example below shows the minimum required properties
```
Remove this comment, or at least make it more accurate to the current code, not the template

Nava2 commented on build.gradle:
```diff
@@ -59,27 +117,27 @@ uploadArchives {
+    tags = ['umple']
+
+    plugins {
+        greetingsPlugin {
```
Don't think this should be `greetingsPlugin` anymore.

Nava2 replied:
@AdamBJ Fix this one..

Nava2 commented on src/integrationTest/groovy/CompilationAndGeneration.groovy:
```diff
@@ -0,0 +1,110 @@
+    private static final Map<String, ArrayList<String>> DIRECTORIES_TO_CHECK
+    // Key = path to directory we want to check, value = the files we expect to find in the directory
+    static {
+        Map<String, ArrayList<String>> aMap = new HashMap<String, ArrayList<String>>()
```
Could you replace this with a [`Multimap`](https://google.github.io/guava/releases/snapshot/api/docs/com/google/common/collect/Multimap.html). You probably want `ArrayListMultimap`.

AdamBJ replied:
I haven't been able to figure out how to make an immutable `ArrayListMultimap`. What I've got right now probably isn't what you're looking for:

```
private static final ImmutableMultimap<String, ArrayList<String>> DIRECTORIES_TO_CHECK
    // Key = path to directory we want to check, value = the files we expect to find in the directory   
    static {
        ImmutableMultimap.Builder<String, ArrayList<String>> builder = ImmutableMultimap.builder();
	    builder.put(DIRECTORIES_ROOT + "subproj/build/classes/generatedSource", ["Subproj.class", "Subproj2.class"])
        builder.put(DIRECTORIES_ROOT + "sub2/build/classes/generatedSource", ["Sub2.class", "Sub22.class"])
        builder.put(DIRECTORIES_ROOT + "subproj/generated/java", ["Subproj.java", "Subproj2.java"])
        builder.put(DIRECTORIES_ROOT + "sub2/generated/java", ["Sub2.java", "Sub22.java"])
	    DIRECTORIES_TO_CHECK = builder.build();
    }
```

I don't think this is all that much of an improvement. Since the map is immutable the usecase for a` multimap` and a `hashmap` don't seem that different.

AdamBJ replied:
I could make it a `<String, String>` and let the multimap handle converting everything to a `Collection<String>` for me when I call `get`, but it seems just as easy to specify it as an `ArrayList` directly.

Nava2 replied:
I like the immutable one more, immutability is always a good thing. 

Also, I don't think you quite understand the multimap, you use `<String, String>` and let it do the `Collection<? extends String>` because it can handle the data better than you by default. 

... I'm more just trying to harass you with Java best practices on this one, there was ultimately nothing wrong with your original one. Immutability is just king (it's why its even in Umple!).

AdamBJ replied:
What's advantages of doing `<String, String>` vs `<String, ArrayList<String>>`? I can see how it's a bit more convenient to simply work with `String`s and let the `Multimap` worry about what collection to prepare for you when you call `get`. Does Guava somehow make a decision about what `Collection<String>` to use based on performance? How does it "handle the data better than you by default"?

Regardless of if we use `<String, String>` or `<String, ArrayList<String>>`, in our case I think it's most convienient to work with `ArrayList<String>` when doing the `put`.

Sure, you could do something like:

```
builder.put(DIRECTORIES_ROOT + "subproj/build/classes/generatedSource", "Subproj.class")
builder.put(DIRECTORIES_ROOT + "subproj/build/classes/generatedSource", "Subproj2.class")
```

And then rely on `Multimap` to create some kind of collection when I call `.get(DIRECTORIES_ROOT + "subproj/build/classes/generatedSource")` But it's fewer lines of code if I add everything at once as an `ArrayList<String>`. And if I do it that way, it makes more sense to me to explicitly specify `<String, ArrayList<String>>` as the type of the `Multimap`.

Nava2 replied:
WFM. 👍

Nava2 commented on src/integrationTest/groovy/CompilationAndGeneration.groovy:
```diff
@@ -0,0 +1,110 @@
+               
+                if (getFileExtension(fileName).equals("java")) {
+                      String actual = SampleFileWriter.readContent(file);
+                      SampleFileWriter.assertFileContent(new File(PATH_TO_TEST_PROJECT + "/verifiedGeneratedOutput", fileName), actual);
```
You could statically import `SampleFileWriter` so this behaves similarly to the JUnit asserts.

Nava2 replied:
@AdamBJ .. and this one..

AdamBJ replied:
I'm getting:

```
C:\Users\i_am_\workspace\umple.gradle\src\integrationTest\groovy\CompilationAndGeneration.groovy: 10: unable to resolve class cruise.umple.util
 @ line 10, column 1.
   import static cruise.umple.util.SampleFileWriter
   ^
```

Why I try a static import. Not sure why. Regardless, I like having the `SampleFileWriter` prefix there, especially for the `assertFileContent` call. Without the prefix someone might think that `assertFileContent` is some kind of JUnit assert function.

Nava2 replied:
You'd do: `import static cruise.umple.util.SampleFileWriter.*`

AdamBJ replied:
Yeah, that works. I still like the idea of having the `SampleFileWriter` sign post, but that's probably because I'm not that familiar with Umple. If you think it's unnecessary I'll use the static import.

AdamBJ replied:
Also, the difference between the `import ...  SampleFileWriter` with and without the `.*` is that without the star we just import the class, and with we import all files inside the `SampleFileWriter` directory?

Nava2 replied:
It's totally a matter of preference, in Umple we statically import it generally. Since we treat it as equivalent to the `Assert.*` functions. 

> Also, the difference between the import ... SampleFileWriter with and without the .* is that without the star we just import the class, and with we import all files inside the SampleFileWriter directory?

Close, the import is importing the _functions_ not the classes.

Nava2 commented on src/integrationTest/groovy/CompilationAndGeneration.groovy:
```diff
@@ -0,0 +1,110 @@
+    }
+    
+    private void clean(final String DIR_PATH) {
+        try {
```
Use [`SampleFileWriter#destroy(String...)`](https://github.com/umple/umple/blob/d9110a6a5064465b77092c34429a44ca84eb7ae4/cruise.umple/src/util/SampleFileWriter_Code.ump#L292)

Nava2 replied:
@AdamBJ ... and lastly this one, then we can merge it. 👍

AdamBJ replied:
I have to get rid of my beautiful clean function 😢 ? Haha

Nava2 commented on src/integrationTest/groovy/CompilationAndGeneration.groovy:
```diff
@@ -0,0 +1,110 @@
+        
+    @Test
+    public void verifyGeneratedAndCompiledFiles() { 
+        for (Map.Entry<String, ArrayList<String>> entry : DIRECTORIES_TO_CHECK.entrySet()) {
```
https://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#walkFileTree(java.nio.file.Path,%20java.nio.file.FileVisitor)

This might be useful.

AdamBJ replied:
Seems pretty neat. What are you suggesting we use it for exactly? Going through the 4 directories to check the contents? The overhead of implementing a class that extends `SimpleFileVisitor` seems unnecessary. We'd need to overwrite `preVisitDirectory` to make sure we only visit the 4 directories we want, and we'd need to overwrite `visitFile` to do the check, right? Seems like a lot of work.

Nava2 commented on src/integrationTest/groovy/CompilationAndGeneration.groovy:
```diff
@@ -0,0 +1,110 @@
+        return false
+    }
+    
+    private String getFileExtension(String fileName) {
```
I find it shocking that this isn't built into the JDK. Huh. It's not, I looked.

Nava2 reviewed (changes requested):
Couple of small changes, then its good from me

AdamBJ committed:
Address comments on PR #11

- Replaced HashMap with ImmutableMultiMap
- Statically imported SampleFileWriter in CompilationAndGeneration.groovy
- Removed the clean method definition, using SampleFileWriter's destroy method instead.
- Minor changes to comments to remove boilerplate

AdamBJ commented:
Just pushed changes that address all your comments except for the potential use of `walkFileTree`. See my comment on that point.

Nava2 commented:
How likely is it that we get folders/files we aren't expecting? (regarding the walkFile)

AdamBJ commented:
We'd start from `...umple.gradle\src\integrationTest\resources\TestProject`.  We wouldn't want to delete anything in that directory, rather we'd want to get into `subproj` and `subproj2`. From there, we *could* delete everything we find, I guess. But it would be safer to only delete the contents of `\sub2\build\classes`, `\subproj\build\classes`, `\sub2\generated`, and `\subproj\generated`. To do it that way there would definitely be some `if/else` stuff happening in the `preVisitDirectory` function to make sure we only start deleting thing once we're in the right place. I think what we've got now is clearer.

AdamBJ commented:
We could also say "only delete .class and .java files". But we'd still need conditional logic to make sure that, for example, nothing in the `verifiedGeneratedOutput` folder gets deleted. What's wrong with doing it the current way?

Nava2 commented:
Oh, I just meant for checking if files are valid. The visitor scaled nicely with more. I'm happy with where this is. Going to merge it.

Nava2 commented:
Wait, why is Travis failing? If it's not related (that is, it's not actually supposed to be passing), I'll merge it.

AdamBJ commented:
It's not related. I enabled Travis last night to see what happened. It fails because there's no `.travis.yml` file yet. Looking into that now

Nava2 commented:
Won't let me merge on mobile. I'll do it when I'm home.

AdamBJ commented:
I can merge it. I think. At least, there's a merge pull request button

Nava2 commented:
:+1:
